File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/NutrimentListItem.java
package openfoodfacts.github.scrachx.openfood.models;

import org.apache.commons.lang.StringUtils;

import static openfoodfacts.github.scrachx.openfood.utils.Utils.getRoundNumber;

public class NutrimentListItem {
    private final boolean displayVolumeHeader;
    private final String modifier;
    private final String servingValue;
    private final String title;
    private final String unit;
    private final String value;

    public NutrimentListItem(boolean displayVolumeHeader) {
        this.displayVolumeHeader = displayVolumeHeader;

        this.title = null;
        this.value = null;
        this.servingValue = null;
        this.unit = null;
        this.modifier = null;
    }

    /**
     * Use a round value for value and servingValue parameters
     *
     * @param title name of nutriment
     * @param value value of nutriment per 100g
     * @param servingValue value of nutriment per serving
     * @param unit unit of nutriment
     * @param modifier one of the following: "<", ">", or "~"
     */
    public NutrimentListItem(String title, String value, String servingValue, String unit,
                             String modifier) {
        this.title = title;
        this.value = getRoundNumber(value);
        this.servingValue = StringUtils.isBlank(servingValue) ? StringUtils.EMPTY : getRoundNumber(servingValue);
        this.unit = unit;
        this.modifier = modifier;

        this.displayVolumeHeader = false;
    }

    public boolean shouldDisplayVolumeHeader() {
        return displayVolumeHeader;
    }

    public CharSequence getTitle() {
        return title;
    }

    public CharSequence getValue() {
        return value;
    }

    /**
     * Get the modifier for this nutriment.
     *
     * @return one of the following: "<", ">", "~", ""
     */
    public CharSequence getModifier() {
        return modifier == null ? "" : modifier;
    }

    public CharSequence getUnit() {
        return unit;
    }

    public CharSequence getServingValue() {
        return servingValue;
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/NutrimentListItem.kt
package openfoodfacts.github.scrachx.openfood.models

import androidx.core.text.bold
import androidx.core.text.buildSpannedString
import openfoodfacts.github.scrachx.openfood.models.Modifier.Companion.DEFAULT
import openfoodfacts.github.scrachx.openfood.utils.getRoundNumber

/**
 * Use a round value for value and servingValue parameters
 *
 * @param title name of nutriment
 * @param valueStr value of nutriment per 100g
 * @param servingValueStr value of nutriment per serving
 * @param unitStr unit of nutriment
 * @param modifierStr one of the following: "<", ">", or "~"
 */
data class NutrimentListItem(
    val title: CharSequence?,
    val valueStr: CharSequence?,
    val servingValueStr: CharSequence?,
    val unitStr: CharSequence?,
    val modifierStr: CharSequence?,
    val displayVolumeHeader: Boolean = false,
) {

    constructor(volumeHeader: Boolean) : this(
        null,
        null,
        null,
        null,
        null,
        volumeHeader
    )

    constructor(title: CharSequence) : this(
        title,
        null,
        null,
        null,
        null
    )

    constructor(
        title: CharSequence,
        nutriment: ProductNutriments.ProductNutriment,
    ) : this(
        title,
        nutriment.per100gInUnit?.value,
        nutriment.perServingInUnit?.value,
        nutriment.unit,
        nutriment.modifier
    )

    constructor(
        title: CharSequence?,
        value: Float?,
        servingValue: Float?,
        unit: MeasurementUnit,
        modifier: Modifier,
        displayVolumeHeader: Boolean = false,
    ) : this(
        title,
        value?.let { getRoundNumber(it) },
        servingValue?.let { getRoundNumber(it) },
        unit.sym,
        modifier.takeUnless { it == DEFAULT }?.sym ?: "",
        displayVolumeHeader
    )
}

fun NutrimentListItem.bold() = copy(
    title = title?.let { bold(it) },
    valueStr = valueStr?.let { bold(it) },
    servingValueStr = servingValueStr?.let { bold(it) },
    unitStr = unitStr?.let { bold(it) },
    modifierStr = modifierStr?.let { bold(it) },
)

private fun bold(msg: CharSequence) = buildSpannedString { bold { append(msg) } }